### PR TITLE
Internal gradcheck wrapper in testing._internal that sets certain flags to True

### LINKED ISSUE #49409: Add global gradcheck setting
## 🚀 Feature

There as two main use cases for gradcheck:
- User facing API to help them check their custom autograd
- Our own testing that use it to check that all our submodules work well with autograd

The problem that we encounter today while adding new features such as vmap and forward AD is that while we want the user facing API to remain stable to avoid breaking user code, we would like the internal testing to test these by default to make sure that functions we provide work with all these gradient-computing features.

## Motivation

We want to avoid having to modify each call to gradcheck in our test suite when we add a new feature.
We also want to make sure new test writer get this by default and it is hard for them to forget to test these.

## Pitch

Provide a global config for gradcheck that allows to change the default to the `gradcheck` and `gradgradcheck` functions.
It will consist in:
- Add a config to gradcheck submodule with all the relevant settings.
- Add API to set these settings easily for test. Also a decorator-based setting to disable them (to easily disable some settings for a single test).
- Changing the API of gradcheck and gradgradcheck to take `None` as default that will translate to whatever is currently in the global config if no value is passed.
- Update our default Test class to set the right setting on setUp and restore the current state in onExit.

## Alternatives

Alternative would be to change all the calls we have in our test suite, but that wouldn't cover new tests being written.
Changing the default of all gradcheck is possible as well but would lead to BC-breaking behavior for users which is not great.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes #49409

There are many call sites where, gradcheck/gradgradcheck is now being implicitly invoked with `check_batched_grad` as True, but they were previously False. Cases fall into two basic categories:
1) the call site was previously using `torch.autograd.gradcheck` but is now changed to use the globally imported function instead
3) the call site was already using globally imported function, but does not explicitly pass `check_batched_grad` flag

Only in the _assertGradAndGradgradChecks cases, which are infrequent, I assumed that the the author is aware that omitting the flag means not applying check_batched_grad=True. (but maybe that is not the case?)

Overall this PR in its current state assumes that unless the author explicitly specified `check_batched_grad=False`, they were just probably not aware of this flag and did not mean to have this flag as False. 

So far exceptions to the above (as discovered by CI) include:
 - Mkldnn (opaque tensors do not have strides) https://app.circleci.com/pipelines/github/pytorch/pytorch/264416/workflows/e4d87886-6247-4305-8526-2696130aa9a4/jobs/10401882/tests
 - all cases in test_sparse (https://app.circleci.com/pipelines/github/pytorch/pytorch/264553/workflows/3c1cbe30-830d-4acd-b240-38d833dccd9b/jobs/10407103)
 - all cases in test_overrides (https://app.circleci.com/pipelines/github/pytorch/pytorch/264553/workflows/3c1cbe30-830d-4acd-b240-38d833dccd9b/jobs/10407236)
 - test_autograd (test_LSTM_grad_and_gradgrad) - (https://app.circleci.com/pipelines/github/pytorch/pytorch/264553/workflows/3c1cbe30-830d-4acd-b240-38d833dccd9b/jobs/10407235)
 - test_data_parallel (test_data_parallel_buffers_requiring_grad) - *SIGSEGV* (https://app.circleci.com/pipelines/github/pytorch/pytorch/264820/workflows/14d89503-040d-4e3d-9f7b-0bc04833589b/jobs/10422697)
 - test_nn (https://app.circleci.com/pipelines/github/pytorch/pytorch/264919/workflows/df79e3ed-8a31-4a8e-b584-858ee99686ff/jobs/10427315)

Possible TODO is to prevent new tests from invoking external gradcheck.

### CODE REVIEW

**Comment by zou3519** [comment]
> So far exceptions to the above (as discovered by CI) include:

@soulitzer when you were looking at the exceptions, did any of them result in silently correct behavior or were all of them loud errors?

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -1894,11 +1892,34 @@ def __enter__(self):
     def __exit__(self, *args):
         pass
 
-def _assertGradAndGradgradChecks(test_case, apply_fn, inputs, check_batched_grad=False):
+
+def gradcheck(fn, inputs, **kwargs):
+    # Wrapper around gradcheck that enables certain keys by default
```
Comment (zou3519) [inline torch/testing/_internal/common_utils.py:1897]: nit: It may be helpful to the reader to provide more context about why we want this. E.g. the reason behind having a testing-internal version of gradcheck is because autograd.gradcheck is a public-facing API. All PyTorch devs doing testing should use the testing-internal version of gradcheck instead of autograd.gradcheck

**Review by zou3519 — COMMENTED** [review]
This looks reasonable to me. I had one high-level question:

batched grad is not going to work for mkldnn, sparse, overrides, or data_parallel anytime soon. Should we disable batched grad testing in those files (via something like `gradcheck = functools.partial(gradcheck, check_batched_grad=False)`?

> Possible TODO is to prevent new tests from invoking external gradcheck.

We should definitely add a lint rule before closing out https://github.com/pytorch/pytorch/issues/49409. It doesn't have to happen in this PR

File: test/test_sparse.py [hunk 2]
```diff
@@ -289,7 +288,7 @@ def test_tensor(x, res):
             def fn(x):
                 return x.to_dense()
             x.requires_grad_(True)
-            gradcheck(fn, (x,), check_sparse_nnz=True)
+            gradcheck(fn, (x,), check_sparse_nnz=True, check_batched_grad=False)
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_sparse.py:291]: Good idea! I don't think it would be too confusing, since the entire thing is designed to be used transparently. OTOH, seems like check_sparse_nnz should be True by default for the file as well but factoring that out into a `functools.partial` might be confusing.

**Comment by soulitzer** [comment]
@zou3519 
All the tests that failed so far generally had errors that mentioned that it was vmap related or had the nice paragraph on how to fix the issue by disabling check_batached_grad. All except for test_data_parallel (test_data_parallel_buffers_requiring_grad), which failed due to a seg fault. You can check [here](https://app.circleci.com/pipelines/github/pytorch/pytorch/264820/workflows/14d89503-040d-4e3d-9f7b-0bc04833589b/jobs/10422697), but its pretty uninformative

Haven't ran the slow tests yet though, do you know how I can tell CI to also run slow tests?

**Comment by zou3519** [comment]
> All the tests that failed so far generally had errors that mentioned that it was vmap related or had the nice paragraph on how to fix the issue by disabling check_batached_grad. All except for test_data_parallel (test_data_parallel_buffers_requiring_grad), which failed due to a seg fault. You can check [here](https://app.circleci.com/pipelines/github/pytorch/pytorch/264820/workflows/14d89503-040d-4e3d-9f7b-0bc04833589b/jobs/10422697), but its pretty uninformative

Thanks for the confirmation!

> > batched grad is not going to work for mkldnn, sparse, overrides, or data_parallel anytime soon. Should we disable batched grad testing in those files (via something like gradcheck = functools.partial(gradcheck, check_batched_grad=False)?
> 
> I think this makes sense to do. There shouldn't be any risk of someone adding non-related tests to any of these files that might actually need batched grad testing right?

Yeah, it seems fine to disable batched grad testing for those files. I don't think someone can write a batched grad test with a sparse tensor / mkldnn tensor / data parallel that succeeds.

**Comment by soulitzer** [comment]
@zou3519, I've disabled batched grads for sparse, data_parallel, and mkldnn. I've also ran the slow tests here (https://github.com/pytorch/pytorch/pull/51295), and I think they should all be passing as well.

I think its ready for another look!

**Review by zou3519 — APPROVED** [review]
lgtm
